x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,363 @@
+    }
+
+    private SparseVectorQueryBuilder(SparseVectorQueryBuilder other, SetOnce<TextExpansionResults> weightedTokensSupplier) {
+        this.fieldName = other.fieldName;
+        this.shouldPruneTokens = other.shouldPruneTokens;
+        this.queryVectors = other.queryVectors;
+        this.inferenceId = other.inferenceId;
+        this.query = other.query;
+        this.tokenPruningConfig = other.tokenPruningConfig;
+        this.weightedTokensSupplier = weightedTokensSupplier;
+    }
+
+    public String getFieldName() {
+        return fieldName;
+    }
+
+    public List<WeightedToken> getQueryVectors() {
+        return queryVectors;
+    }
+
+    public boolean shouldPruneTokens() {
+        return shouldPruneTokens;
+    }
+
+    public TokenPruningConfig getTokenPruningConfig() {
+        return tokenPruningConfig;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        if (weightedTokensSupplier != null) {
+            throw new IllegalStateException("weighted tokens supplier must be null, can't serialize suppliers, missing a rewriteAndFetch?");
+        }
+
+        out.writeString(fieldName);
+        out.writeBoolean(shouldPruneTokens);
+        out.writeOptionalCollection(queryVectors);
+        out.writeOptionalString(inferenceId);
+        out.writeOptionalString(query);
+        out.writeOptionalWriteable(tokenPruningConfig);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(FIELD_FIELD.getPreferredName(), fieldName);
+        if (queryVectors != null) {
+            builder.startObject(QUERY_VECTOR_FIELD.getPreferredName());
+            for (var token : queryVectors) {
+                token.toXContent(builder, params);
+            }
+            builder.endObject();
+        } else {
+            builder.field(INFERENCE_ID_FIELD.getPreferredName(), inferenceId);
+            builder.field(QUERY_FIELD.getPreferredName(), query);
+        }
+        builder.field(PRUNE_FIELD.getPreferredName(), shouldPruneTokens);
+        if (tokenPruningConfig != null) {
+            builder.field(PRUNING_CONFIG_FIELD.getPreferredName(), tokenPruningConfig);
+        }
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        if (queryVectors == null) {
+            return new MatchNoDocsQuery("Empty query vectors");
+        }
+
+        final MappedFieldType ft = context.getFieldType(fieldName);
+        if (ft == null) {
+            return new MatchNoDocsQuery("The \"" + getName() + "\" query is against a field that does not exist");
+        }
+
+        final String fieldTypeName = ft.typeName();
+        if (fieldTypeName.equals(ALLOWED_FIELD_TYPE) == false) {
+            throw new ElasticsearchParseException(
```

Comment:
should we return here `IllegalArgumentException` instead of `ElasticsearchParseException`. I interpret `ElasticsearchParseException` as syntax parsing exception.

Reply:
I'll change, thanks!